### PR TITLE
fix(gemma3,4): move the padding logic from model to scheduler

### PR DESCRIPTION
**Pull request by rebel-eunji**
### 🚀 Summary of Changes

1. Remove the padding logic in the processor.
2. Calculate the padding size in the scheduler considering prefill chunk size and image_prefill_chunk_size.
3. Sync vllm and optimum for compilation - prefill_chunk_size, image_prefill_chunk_size
    - parameterize the prefill_chunk_size when we compile the model
    - load the prefill chunk size from pre-compiled model
4. cache the padding size for each request to reuse it

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
5. Verify output: `...`
6. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/model_executor/models/optimum/compilation/__init__.py [hunk 1]
```diff
@@ -82,11 +82,14 @@ def for_architecture(
...
         num_devices: int,
+        prefill_chunk_size: int | None = None,
         rbln_overrides: dict[str, Any] | None = None,
     ) -> "RBLNCompileSpec":
         """Build a compile spec from vllm-rbln inputs, dispatched by architecture."""
         if is_generation_arch(config):
-            spec = cls._for_decoder(batch_size, block_size, max_model_len, num_devices)
+            spec = cls._for_decoder(
+                batch_size, block_size, max_model_len, num_devices, prefill_chunk_size
+            )
```
Comment (rebel-thkim) [inline vllm_rbln/model_executor/models/optimum/compilation/__init__.py:92]: Is it necessary for `decoderonly` as well?

File: vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py [hunk 2]
```diff
@@ -303,21 +303,15 @@ def preprocess(
         padded_local_block_table_ids = self.pad_to_2d(
             local_block_table_ids, decoder_batch_size, 1, pad_value, torch.int16
         )
-        padded_pad_len = self.pad_to_2d(pad_lens, decoder_batch_size, 1, 0)
+        # padded_pad_len = self.pad_to_2d(pad_lens, decoder_batch_size, 1, 0)
```
Comment (rebel-thkim) [inline vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py]: "erase" seems better than "comment."

File: vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py [hunk 3]
```diff
@@ -303,21 +303,15 @@ def preprocess(
...
-
-        # cache_positions:
-        #  the index including padding between text and image
-        # pad_lens:
-        #   the size of padding
-        # position_ids:
-        #   the index of the token to be decoded in the sequence.
-        position_ids = padded_cache_positions - padded_pad_len
+        # FIXME remove the logic
+        position_ids = padded_cache_positions
```
Comment (rebel-thkim) [inline vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py:314]: Is this because the scheduler takes the padding between blocks into account when passing the `cache_position`?

File: vllm_rbln/model_executor/models/optimum/gemma3.py [hunk 4]
```diff
@@ -91,119 +67,11 @@ class RBLNChunkedPrefillPadMixin(_ProcessorBase):
 
     # MRO note: mix in BEFORE the HF ``*MultiModalProcessor`` so this ``apply``
     # wraps theirs (``super().apply`` resolves to the HF processor).
+    #
```
Comment (rebel-thkim) [inline vllm_rbln/model_executor/models/optimum/gemma3.py]: erase '#'

File: vllm_rbln/model_executor/models/optimum/gemma3.py [hunk 5]
```diff
@@ -48,28 +46,6 @@
...
-    """
-    n = len(token_types)
-    end = min(start + cap, n)
-    i = start
-    while i < end and token_types[i] == value:
-        i += 1
-    return i - start
-
-
 class RBLNChunkedPrefillPadMixin(_ProcessorBase):
```
Comment (rebel-thkim) [inline vllm_rbln/model_executor/models/optimum/gemma3.py]: This class no longer seems necessary.

File: vllm_rbln/model_executor/models/optimum/compilation/__init__.py (hunk 1 above)
Comment (rebel-eunji) [reply vllm_rbln/model_executor/models/optimum/compilation/__init__.py:92]: It is not really related with this pr, but prefill_chunk_size is also important in decoder-only models for prefix caching.

File: vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py [hunk 6]
```diff
@@ -303,21 +303,15 @@ def preprocess(
         padded_local_block_table_ids = self.pad_to_2d(
             local_block_table_ids, decoder_batch_size, 1, pad_value, torch.int16
         )
-        padded_pad_len = self.pad_to_2d(pad_lens, decoder_batch_size, 1, 0)
+        # padded_pad_len = self.pad_to_2d(pad_lens, decoder_batch_size, 1, 0)
         padded_cache_positions = self.pad_to_2d(
             cache_positions, decoder_batch_size, 1, 0
         )
         padded_attention_mask = self.pad_to_2d(
```
Comment (rebel-eunji) [inline vllm_rbln/model_executor/models/optimum/optimum_attention/optimum_attention_strategy.py:310]: @rebel-thkim attention_mask is also unused from now on, right?

File: vllm_rbln/v1/core/optimum_kv_cache_manager.py [hunk 7]
```diff
@@ -103,6 +123,99 @@ def __init__(
...
+            assert self.prefill_chunk_size is not None, (
+                "prefill_chunk_size must be set when image_prefill_chunk_sizes is empty"
+            )
+            return self.prefill_chunk_size
+        fits = [b for b in buckets if b >= run_len]
+        return min(fits) if fits else max(buckets)
+
+    def _chunked_prefill_pad(self, request: Request, query_len: int) -> int:
+        # FIXME chunk size?????
+        text_chunk = self.prefill_chunk_size
```
Comment (rebel-eunji) [inline vllm_rbln/v1/core/optimum_kv_cache_manager.py:184]: @rebel-dkhong 
please make sure the chunk size setting. As I know the image chunk size is dependent on the image size, not request. It's hard to choose only one in this logic.

File: vllm_rbln/v1/core/optimum_kv_cache_manager.py [hunk 8]
```diff
@@ -103,6 +123,99 @@ def __init__(
...
+        (e.g. text-only), or the largest bucket when the run exceeds all of them.
+        """
+        buckets = self.image_prefill_chunk_sizes
+        if not buckets:
+            assert self.prefill_chunk_size is not None, (
+                "prefill_chunk_size must be set when image_prefill_chunk_sizes is empty"
+            )
+            return self.prefill_chunk_size
+        fits = [b for b in buckets if b >= run_len]
+        return min(fits) if fits else max(buckets)
```
Comment (rebel-dkhong) [inline vllm_rbln/v1/core/optimum_kv_cache_manager.py]: Please raise error if `fits` is empty.

File: tests/v1/core/test_chunked_prefill_pad.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
+# Copyright 2025 Rebellions Inc. All rights reserved.
```
Comment (rebel-eunji) [inline tests/v1/core/test_chunked_prefill_pad.py]:
```suggestion
# Copyright 2026 Rebellions Inc. All rights reserved.
```

**Review by rebel-thkim — APPROVED** [review]
👍
